allow to override prefers-color-scheme via url param for guests

jonathanmmm opened the pull request:
using ?force-color-scheme=light should make the theme light and ?force-color-scheme=dark should make it dark.
Useful for iframes to match a website theme and to send a specific theme to a guest.

I couldn't test it locally, as it seems that javascript is somehow like compiled (and the src folder is missing for example) and have no developer setup. If needed, I can set one up.

My first pull request to nextcloud, if you have feedback how to make it better next time, tell me.

jonathanmmm committed:
allow to override prefers-color-scheme via url param for guests

using ?force-color-scheme=light should make the theme light and ?force-color-scheme=dark should make it dark.
Useful for iframes to match a website theme and to send a specific theme to a guest.

Signed-off-by: jonathanmmm <32403139+jonathanmmm@users.noreply.github.com>

jonathanmmm commented:
Could solve this issue: https://github.com/nextcloud/server/issues/29699

szaimen commented:
/compile amend /

jonathanmmm commented:
Ok, somehow it doesn't do anything.
Seems to be that /server/apps/accessibility/lib/AppInfo/Application.php
`$userValues = ['dark'];

			$hash = md5(implode('-', $userValues));
			$linkToCSS = $urlGenerator->linkToRoute(self::APP_ID . '.accessibility.getCss', ['md5' => $hash]);
			\OCP\Util::addHeader('link', ['rel' => 'stylesheet', 'media' => '(prefers-color-scheme: dark)', 'href' => $linkToCSS]);
			\OCP\Util::addScript('accessibility', 'accessibilityoca');`

If I put into the third line `light` instead of `dark` it changes to light theme.
Testing in incognito mode.

Does somebody know, if `accessibilityoca.js` in src has to be compiled and lands then in the js folder?

Don't know how these js files with the super long lines work, are they compiled or do I have to change them too?

This pull and commit seem to made prefer-color-scheme possible in the first place https://github.com/nextcloud/server/pull/21366/files
https://github.com/nextcloud/server/commit/ede5db10c3d5636305d5e9f5a9b41df8741d3e85

CarlSchwan commented on apps/accessibility/src/accessibilityoca.js:
```diff
@@ -24,8 +24,19 @@
-if (OCA.Accessibility.checkMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
-	// Overwrite the theme for Guests based on the prefers-color-scheme
+// Force a specific theme for guests based on force-color-scheme param
+var srv_url = new URL(window.location.href)
+var force_color_scheme = srv_url.searchParams.get("force-color-scheme")
+if(force_color_scheme !== null) {
+	if(force_color_scheme=="dark") {
+		OCA.Accessibility.theme = 'dark'
+	}
+	else {
+		OCA.Accessibility.theme = 'light'
+	}
+}
+else if (OCA.Accessibility.checkMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
```
```suggestion
const srvUrl = new URL(window.location.href)
const forceColorScheme = srvUrl.searchParams.get("force-color-scheme")
if (forceColorScheme !== null && (forceColorScheme === 'dark' || forceColorScheme === 'light')) {
    OCA.Accessibility.theme = forceColorScheme;
} else if (OCA.Accessibility.checkMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
```

A bit nicer :)

jonathanmmm replied:
You are right, could you push it to my branch (have given access to reviewers/maintainers)? Or should I do it?

CarlSchwan committed:
Update apps/accessibility/src/accessibilityoca.js

jonathanmmm committed:
fixed string in double quote to single quote

jonathanmmm commented:
@CarlSchwan If you have motivation+time, we could takle this feature together.

jonathanmmm commented:
Research:

https://github.com/nextcloud/server/commit/291640273dfc17aa240a48d7d4a1553d76d54e41
override these changes in the commit if url parameter is given

and change it so that line 27-30 the value in line 29 is not changed if url parameter is given
https://github.com/nextcloud/server/blob/master/apps/accessibility/src/accessibilityoca.js

jonathanmmm committed:
works with dark mode enabled in system (not with light mode and forcing dark mode)

jonathanmmm commented:
@CarlSchwan It is working now when dark mode is enabled in the system (forcing light mode), but not the other way around and I don't know. I even tried and replaced every `theme--light` string I could find under accessibility and replaced it with `theme--dark`, still light mode.

skjnldsv commented:
As there is no feedback since a while I will close this ticket. 
If you will decide to work on this feature again and if it hasn't been fixed or implemented already, feel free to re-open and solve the various conflicts.

Thanks for the interest in Nextcloud and the effort put into this! 🙇
